File: java-core/src/main/java/org/maw/wedding/navigation/EventBus.java
package org.maw.wedding.navigation;

import java.util.ArrayList;
import java.util.List;

public class EventBus implements Bus {

    private List<NavigationListener> mListenerList = new ArrayList<>();

    @Override
    public void subscibe(NavigationListener listener) {
        mListenerList.add(listener);
    }

    @Override
    public void unSubscribe(NavigationListener listener) {
        mListenerList.remove(listener);
    }

    @Override
    public void post(NavigationEvent navigationEvent) {
        for (NavigationListener listener : mListenerList) {
            listener.onNavigationEvent(navigationEvent);
        }
    }
}

File: kotlin-core/src/main/kotlin/org/maw/wedding/navigation/EventBus.kt
package org.maw.wedding.navigation

import java.util.*

class EventBus : Bus {

    private val mListenerMap = HashMap<Bus.EVENT_TYPE, MutableList<EventListener>>()

    init {
        mListenerMap.put(Bus.EVENT_TYPE.NAVIGATION, ArrayList<EventListener>())
        mListenerMap.put(Bus.EVENT_TYPE.OTHER, ArrayList<EventListener>())
    }

    override fun subscibe(eventType: Bus.EVENT_TYPE, listener: EventListener) {
        mListenerMap.get(eventType)!!.add(listener)
    }

    override fun unSubscribe(eventType: Bus.EVENT_TYPE, listener: EventListener) {
        mListenerMap.get(eventType)!!.remove(listener)
    }

    override fun post(event: Event) {
        for (listener in mListenerMap[event.getEventType()]!!) {
            listener.onEvent(event)
        }
    }
}
